### ISSUE TITLE
Redirects for autocomplete

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#1380]
Right now, the autocomplete list is populated based on a weighted edit distance between what's been typed already and the symbol being considered.  This works well when a student knows part or all of the function name they are looking for, but it misses cases where a student wants to draw a square, or a triangle.  There is no `square` function, but squares can be drawn using `rectangle`.  There is no `triangle`, but triangles can be drawn using `polygon`.

I propose to add this knowledge in the form of redirects for autocomplete, so that if a student triggers autocomplete with something close to `square`, we should suggest `rectangle` as a completion.  Similarly, `triangle` should redirect to `polygon`.  Maybe `ellipse` and `oval` should redirect to `scaled(circle`?  Though it's weird to add unbalanced parentheses.

**Comment by Powell-v2** [comment]
Should these suggestions go through the sieving process or I can just stick them into the final list?

**Comment by cdsmith** [comment]
I'm not sure what you mean by the sieving process.  Referring to the code at https://github.com/google/codeworld/blob/e6fa062a3c1c9f0895ab65fe32eefba03d1ddf1e/web/js/codeworld_shared.js#L487, I think they should go through the same scoring and sorting process as the existing suggestions, except that they should only be added to tier 3 (the fuzzy matching tier) so that they are never chosen automatically.  The scoring should be done on the trigger word (e.g., `square` or `triangle`, and not on the word that will actually be inserted.

It's very likely that the scores should be demoted somehow, perhaps even by a different amount for each one (so, for instance, `square` should maybe redirect to both `rectangle` and `polygon`, but `polygon` should be demoted further), but that can be added after the initial implementation.

It would also be nice to add something to the hint doc to explain why you're getting this suggestion.  But again, don't include that in the initial implementation.  Let's get it working first, and then make the UI changes afterward.

**Comment by Powell-v2** [comment]
Please share your thoughts on the initial implementation.

**Comment by Powell-v2** [comment]
> It would also be nice to add something to the hint doc to explain why you're getting this suggestion.

Here's one way to go about it:

<img width="934" alt="Screenshot 2020-07-27 at 14 21 19" src="https://user-images.githubusercontent.com/25308326/88541327-d9c96800-d014-11ea-84d2-e6a853076cd7.png">

If you prefer it done differently, I can come up with few other ideas.

What should this note actually say? I was thinking that it might be something along the lines of:

_"If you are looking for {search_term}, it doesn't exist. Try this symbol instead."_

Also, do you think it's a good idea to highlight these replacements somehow within the list?

**Comment by cdsmith** [comment]
I left comments on the pull request about this.  I didn't think about this until now, but we're already adding all possible replacements into the list.  I actually now don't think we want to add any new entries to autocomplete at all.  Surprisingly (at least to me), this is just going to be a scoring change!

Still, we can add the rest of these features.  The new score will be the best of the organic score for each candidate, and the scores that go via replacements.  So we can still observe whether the best score was obtained via a replacement, and do any highlighting or indentation.  My instinct is to say not to highlight them, but to add an annotation to the doc dialog.

As far as the text, the gold standard would be to include that in the replacement table, so that when we add redirects into the table, the text can be customized.  For `triangle`, it might say "A triangle is a polygon with three vertices."  Sometimes the redirect might just be for a common misspelling or alternate spelling, so the explanation text should be optional, and nothing changed if it's missing.

**Comment by Powell-v2** [comment]
> Surprisingly (at least to me), this is just going to be a scoring change

That's true! I didn't realize this either, until now. This brings up a question, though - what to do with [direct hits](https://github.com/google/codeworld/blob/master/web/js/codeworld_shared.js#L529)?

> the replacement table

What is this, exactly?

**Comment by cdsmith** [comment]
> direct hits

Technically, we should only be scoring entries in `found[2]` using this logic.  But I don't think that it matters, honestly, because in practice an item in tiers 1 or 2 should always score higher relative to its true completion than a redirect, especially after the penalty.  So I don't think a change is needed there.

> the replacement table

I was referring to this dictionary: https://github.com/google/codeworld/pull/1400/files#diff-908d29aadd4abb809c5cd7ecafc602c8R540

**Comment by cdsmith** [comment]
Here's a more complete set of redirects:

* `positive` -> `abs`
* `magnitude` -> `abs`
* `arccos` -> `acos`
* `interactionOf` -> `activityOf` (from a deprecation)
* `semicircle` -> `arc`
* `arcsin` -> `asin`
* `arctan` -> `atan`
* `blend` -> `mixed`
* `colors` -> `assortedColors`
* `loop` -> `closedCurve`
* `cosine` -> `cos`
* `text` -> `lettering`
* `MouseMovement` -> `PointerMovement` (from a deprecation)
* `MousePress` -> `PointerPress` (from a deprecation)
* `MouseClick` -> `PointerPress`
* `PointerClick` -> `PointerPress`
* `MouseRelease` -> `PointerRelease` (from a deprecation)
* `triangle` -> `polygon`
* `quadrilateral` -> `polygon`
* `parallelogram` -> `polygon`
* `trapezoid` -> `polygon`
* `pentagon` -> `polygon`
* `hexagon` -> `polygon`
* `septagon` -> `polygon`
* `octagon` -> `polygon`
* `line` -> `polyline`
* `main` -> `program` (from a deprecation)
* `square` -> `rectangle`
* `turn` -> `rotated`
* `stretch` -> `scaled`
* `reflect` -> `scaled`
* `solidArc` -> `sector`
* `sine` -> `sin`
* `solidCurve` -> `solidClosedCurve`
* `solidLoop` -> `solidClosedCurve`
* `solidTriangle` -> `solidPolygon`
* `solidQuadrilateral` -> `solidPolygon`
* `solidParallelogram` -> `solidPolygon`
* `solidTrapezoid` -> `solidPolygon`
* `solidPentagon` -> `solidPolygon`
* `solidHexagon` -> `solidPolygon`
* `solidSeptagon` -> `solidPolygon`
* `solidOctagon` -> `solidPolygon`
* `solidSquare` -> `solidRectangle`
* `root` -> `sqrt`
* `radical` -> `sqrt`
* `tangent` -> `tan`
* `thickLoop` -> `thickClosedCurve`
* `thickTriangle` -> `thickPolygon`
* `thickQuadrilateral` -> `thickPolygon`
* `thickParallelogram` -> `thickPolygon`
* `thickTrapezoid` -> `thickPolygon`
* `thickPentagon` -> `thickPolygon`
* `thickHexagon` -> `thickPolygon`
* `thickSeptagon` -> `thickPolygon`
* `thickOctagon` -> `thickPolygon`
* `thickLine` -> `thickPolyline`
* `thickSquare` -> `thickRectangle`
* `move` -> `translated`

I could see this growing much further.  Perhaps it should go in a JSON file, and be loaded at runtime.

**Comment by Powell-v2** [comment]
> what to do with direct hits?

Let me clarify a use case I had in mind when asking this. Imagine someone types 'squa' and invokes autocompletion. As things currently stand, 'squa' will automatically morph into 'squareRoot' as we are emptying the bucket with fuzzy candidates. Do we still want to show potential replacements in such cases?

**Comment by cdsmith** [comment]
Oh, that's actually a really interesting point.  So the autocomplete for square would be useless, because `squareRoot` exists.  Huh.  I will think about this more.

One option would be to just check in advance that none of the redirects are prefixes of existing words.  We can probably fix the `square` situation by just removing `squareRoot`, which I doubt anyone has ever used.  (Honestly, if this feature had existed, it would just be a redirect to `sqrt`.  Same for `absoluteValue`, which should just be a redirect to `abs`)

**Comment by cdsmith** [comment]
I just pushed a commit to remove `squareRoot` and `absoluteValue` from the codeworld-base `Prelude`.  This should resolve the concern about the redirect for `square`.

**Comment by cdsmith** [comment]
This is working, but the redirect table could use some love.  I'll try to do that today.
